CUDA 11 Test: `cupy_tests.fft_tests.test_fft.TestMultiGpu`

```
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_2_{n=None, norm=None, shape=(64,)}::test_fft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_2_{n=None, norm=None, shape=(64,)}::test_ifft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_6_{n=None, norm='ortho', shape=(64,)}::test_fft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_6_{n=None, norm='ortho', shape=(64,)}::test_ifft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_24_{n=64, norm=None, shape=(0,)}::test_fft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_24_{n=64, norm=None, shape=(0,)}::test_ifft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_26_{n=64, norm=None, shape=(64,)}::test_fft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_26_{n=64, norm=None, shape=(64,)}::test_ifft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_28_{n=64, norm='ortho', shape=(0,)}::test_fft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_28_{n=64, norm='ortho', shape=(0,)}::test_ifft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_30_{n=64, norm='ortho', shape=(64,)}::test_fft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuFft_param_30_{n=64, norm='ortho', shape=(64,)}::test_ifft - AssertionError: Parameterized test failed.
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_0_{n=None, norm=None, shape=(64,)}::test_fft - AssertionError: Parameterized test f...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_0_{n=None, norm=None, shape=(64,)}::test_ifft - AssertionError: Parameterized test ...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_1_{n=None, norm=None, shape=(128,)}::test_fft - AssertionError: Parameterized test ...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_1_{n=None, norm=None, shape=(128,)}::test_ifft - AssertionError: Parameterized test...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_2_{n=None, norm='ortho', shape=(64,)}::test_fft - AssertionError: Parameterized tes...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_2_{n=None, norm='ortho', shape=(64,)}::test_ifft - AssertionError: Parameterized te...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_3_{n=None, norm='ortho', shape=(128,)}::test_fft - AssertionError: Parameterized te...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_3_{n=None, norm='ortho', shape=(128,)}::test_ifft - AssertionError: Parameterized t...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_4_{n=64, norm=None, shape=(64,)}::test_fft - AssertionError: Parameterized test fai...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_4_{n=64, norm=None, shape=(64,)}::test_ifft - AssertionError: Parameterized test fa...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_5_{n=64, norm=None, shape=(128,)}::test_fft - AssertionError: Parameterized test fa...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_5_{n=64, norm=None, shape=(128,)}::test_ifft - AssertionError: Parameterized test f...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_6_{n=64, norm='ortho', shape=(64,)}::test_fft - AssertionError: Parameterized test ...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_6_{n=64, norm='ortho', shape=(64,)}::test_ifft - AssertionError: Parameterized test...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_7_{n=64, norm='ortho', shape=(128,)}::test_fft - AssertionError: Parameterized test...
FAILED tests/cupy_tests/fft_tests/test_fft.py::TestMultiGpuPlanCtxManagerFft_param_7_{n=64, norm='ortho', shape=(128,)}::test_ifft - AssertionError: Parameterized tes...
```

--- a/cupy/testing/parameterized.py
+++ b/cupy/testing/parameterized.py
@@ -6,6 +6,7 @@ import typing as tp  # NOQA
 import unittest
 
 from cupy.testing import _bundle
+from cupy.cuda import cufft
 from cupy.cuda import driver
 
 
@@ -91,7 +92,7 @@ def _parameterize_test_case(base, i, param):
                 for k, v in sorted(param.items()):
                     s.write('  {}: {}\n'.format(k, v))
                 err_class = e.__class__
-                if isinstance(e, driver.CUDADriverError):
+                if isinstance(e, (driver.CUDADriverError, cufft.CuFFTError)):
                     err_class, = err_class.__bases__
                 raise err_class(s.getvalue()).with_traceback(e.__traceback__)
         return new_method

--- a/tests/cupy_tests/fft_tests/test_fft.py
+++ b/tests/cupy_tests/fft_tests/test_fft.py
@@ -53,7 +53,7 @@ def multi_gpu_config(gpu_configs=None):
     """Decorator for parameterized tests with different GPU configurations.
 
     Args:
-        gpu_configs(list of list): The GPUs to test.
+        gpu_configs (list of list): The GPUs to test.
 
     .. notes:
         The decorated tests are skipped if no or only one GPU is available.
@@ -71,14 +71,16 @@ def multi_gpu_config(gpu_configs=None):
                         assert nGPUs >= 2, 'Must use at least two gpus'
                         config.use_multi_gpus = True
                         config.set_cufft_gpus(gpus)
+                        self.gpus = gpus
 
                         impl(self, *args, **kw)
-                    except Exception:
+                    except:
                         print('GPU config is:', gpus)
                         raise
             finally:
                 config.use_multi_gpus = use_multi_gpus
                 config._devices = _devices
+                del self.gpus
 
         return test_func
     return decorator
@@ -159,6 +161,16 @@ class TestFftOrder(unittest.TestCase):
         return out
 
 
+def _skip_multi_gpu_bug(shape, gpus):
+    # avoid CUDA 11 bug triggered by
+    # - batch = 1
+    # - gpus = [1, 0]
+    if (cupy.cuda.runtime.runtimeGetVersion() == 11000
+            and len(shape) == 1
+            and gpus == [1, 0]):
+        raise unittest.SkipTest('avoid CUDA 11 bug')
+
+
 # Almost identical to the TestFft class, except that
 # 1. multi-GPU cuFFT is used
 # 2. the tested parameter combinations are adjusted to meet the requirements
@@ -169,11 +181,14 @@ class TestFftOrder(unittest.TestCase):
 }))
 @testing.multi_gpu(2)
 class TestMultiGpuFft(unittest.TestCase):
+
     @multi_gpu_config(gpu_configs=[[0, 1], [1, 0]])
     @testing.for_complex_dtypes()
     @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
                                  contiguous_check=False)
     def test_fft(self, xp, dtype):
+        _skip_multi_gpu_bug(self.shape, self.gpus)
+
         a = testing.shaped_random(self.shape, xp, dtype)
         out = xp.fft.fft(a, n=self.n, norm=self.norm)
 
@@ -191,6 +206,8 @@ class TestMultiGpuFft(unittest.TestCase):
     @testing.with_requires('numpy!=1.17.0')
     @testing.with_requires('numpy!=1.17.1')
     def test_ifft(self, xp, dtype):
+        _skip_multi_gpu_bug(self.shape, self.gpus)
+
         a = testing.shaped_random(self.shape, xp, dtype)
         out = xp.fft.ifft(a, n=self.n, norm=self.norm)
 
@@ -216,6 +233,8 @@ class TestMultiGpuFftOrder(unittest.TestCase):
     @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
                                  contiguous_check=False)
     def test_fft(self, xp, dtype):
+        _skip_multi_gpu_bug(self.shape, self.gpus)
+
         a = testing.shaped_random(self.shape, xp, dtype)
         if self.data_order == 'F':
             a = xp.asfortranarray(a)
@@ -232,6 +251,8 @@ class TestMultiGpuFftOrder(unittest.TestCase):
     @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
                                  contiguous_check=False)
     def test_ifft(self, xp, dtype):
+        _skip_multi_gpu_bug(self.shape, self.gpus)
+
         a = testing.shaped_random(self.shape, xp, dtype)
         if self.data_order == 'F':
             a = xp.asfortranarray(a)
@@ -640,6 +661,8 @@ class TestMultiGpuPlanCtxManagerFft(unittest.TestCase):
     @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
                                  contiguous_check=False)
     def test_fft(self, xp, dtype):
+        _skip_multi_gpu_bug(self.shape, self.gpus)
+
         a = testing.shaped_random(self.shape, xp, dtype)
         if xp is cupy:
             from cupyx.scipy.fftpack import get_fft_plan
@@ -662,6 +685,8 @@ class TestMultiGpuPlanCtxManagerFft(unittest.TestCase):
     @testing.numpy_cupy_allclose(rtol=1e-4, atol=1e-7, accept_error=ValueError,
                                  contiguous_check=False)
     def test_ifft(self, xp, dtype):
+        _skip_multi_gpu_bug(self.shape, self.gpus)
+
         a = testing.shaped_random(self.shape, xp, dtype)
         if xp is cupy:
             from cupyx.scipy.fftpack import get_fft_plan

--- a/tests/cupy_tests/fft_tests/test_fft.py
+++ b/tests/cupy_tests/fft_tests/test_fft.py
@@ -163,10 +163,10 @@ class TestFftOrder(unittest.TestCase):
 
 # See #3757 and NVIDIA internal ticket 3093094
 def _skip_multi_gpu_bug(shape, gpus):
-    # avoid CUDA 11 bug triggered by
+    # avoid CUDA 11.0 (will be fixed by CUDA 11.2) bug triggered by
     # - batch = 1
     # - gpus = [1, 0]
-    if (cupy.cuda.runtime.runtimeGetVersion() == 11000
+    if (11000 <= cupy.cuda.runtime.runtimeGetVersion() < 11200
             and len(shape) == 1
             and gpus == [1, 0]):
         raise unittest.SkipTest('avoid CUDA 11 bug')

--- a/tests/cupy_tests/fft_tests/test_fft.py
+++ b/tests/cupy_tests/fft_tests/test_fft.py
@@ -74,7 +74,7 @@ def multi_gpu_config(gpu_configs=None):
                         self.gpus = gpus
 
                         impl(self, *args, **kw)
-                    except:
+                    except Exception:
                         print('GPU config is:', gpus)
                         raise
             finally:
@@ -161,6 +161,7 @@ class TestFftOrder(unittest.TestCase):
         return out
 
 
+# See #3757 and NVIDIA internal ticket 3093094
 def _skip_multi_gpu_bug(shape, gpus):
     # avoid CUDA 11 bug triggered by
     # - batch = 1

